source/endpoints_eval_util.c

```diff
@@ -0,0 +1,420 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/common/logging.h>
+#include <aws/common/string.h>
+#include <aws/common/json.h>
+#include <aws/sdkutils/private/endpoints_eval_util.h>
+#include <aws/sdkutils/sdkutils.h>
+
+#include <inttypes.h>
+
+#ifdef _MSC_VER
+#    pragma warning(disable : 4204)
+#    pragma warning(disable : 4706)
+#    pragma warning(disable : 4996)
+#endif
+
+/* 4 octets of 3 chars max + 3 separators + null terminator */
+#define AWS_IPV4_STR_LEN 16
+#define IP_CHAR_FMT "%03" SCNu16
+
+/* arbitrary max length of a region. curent longest region name is 16 chars */
+#define AWS_REGION_LEN 50
+
+struct aws_byte_cursor aws_byte_cursor_from_substring(const struct aws_string *src, size_t start, size_t end) {
+    AWS_PRECONDITION(aws_string_is_valid(src));
+    AWS_PRECONDITION(start < end && end <= src->len);
+
+    return aws_byte_cursor_from_array(aws_string_bytes(src) + start, end - start);
+}
+
+bool aws_is_ipv4(struct aws_byte_cursor host) {
+    if (host.len > AWS_IPV4_STR_LEN - 1) {
+        return false;
+    }
+
+    char copy[AWS_IPV4_STR_LEN] = {0};
+    memcpy(copy, host.ptr, host.len);
+
+    uint16_t octet[4] = {0};
+    char remainder[2] = {0};
+    if (4 != sscanf(
+                 copy,
+                 IP_CHAR_FMT "." IP_CHAR_FMT "." IP_CHAR_FMT "." IP_CHAR_FMT "%1s",
+                 &octet[0],
+                 &octet[1],
+                 &octet[2],
+                 &octet[3],
+                 remainder)) {
+        return false;
+    }
+
+    for (size_t i = 0; i < 4; ++i) {
+        if (octet[i] > 255) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool aws_is_ipv6(struct aws_byte_cursor host, bool is_uri_encoded) {
+    if (host.len == 0) {
+        return false;
+    }
+
+    if (is_uri_encoded) {
+        if (host.ptr[0] != '[' || host.ptr[host.len - 1] != ']') {
+            return false;
+        }
+        aws_byte_cursor_advance(&host, 1);
+        --host.len;
```

Comment:
if host.len started as 1, and we take 2 away, then host.len will underflow and flip to SIZE_MAX

Reply:
if check right before will exit if cursor does not start with [ or end with ], so it must have at least 2 chars. point taken though, let me see if i can make it clearer